### PR TITLE
Use bddt to simplify guardt

### LINKED ISSUE #306: Example for improvement of guard simplification
Supporting material for feature request : not bug.

@tautschnig I remember we talked ages ago about improving the simplification of guard expressions (to help a problem Pascal was having with switch).  It seems that:

sv-benchmarks/c/recursive-simple/id_o1000_false-unreach-call.c

(from SV-COMP) gives a good example of guards that look like they should be simplifable.

### PR DESCRIPTION
**Pull request by tautschnig**
This should fix #306

### CODE REVIEW

File: src/goto-symex/symex_goto.cpp [hunk 1]
```diff
@@ -366,11 +367,38 @@ void goto_symext::phi_function(
...
+      // tmp_guard-=dest_state.guard;
+      if(!goto_state.guard_before_branch.is_constant())
+      {
+        assert(tmp_guard2.id()==ID_and);
+
+        if(goto_state.guard_before_branch.id()==ID_and)
+        {
+          assert(tmp_guard2.operands().size()>
+                 goto_state.guard_before_branch.operands().size());
+          for(const auto & op : goto_state.guard_before_branch.operands())
```
Comment (peterschrammel) [inline src/goto-symex/symex_goto.cpp]: auto &op

File: src/goto-symex/symex_goto.cpp [hunk 2]
```diff
@@ -366,11 +367,38 @@ void goto_symext::phi_function(
...
+          {
+            // assert(tmp_guard2.operands().front()==op);
```
Comment (peterschrammel) [inline src/goto-symex/symex_goto.cpp]: use #if 0 or remove

File: src/goto-symex/symex_goto.cpp [hunk 3]
```diff
@@ -366,11 +367,38 @@ void goto_symext::phi_function(
     else
     {
       guardt tmp_guard(goto_state.guard);
+      guardt tmp_guard2(goto_state.guard);
 
       // this gets the diff between the guards
-      tmp_guard-=dest_state.guard;
-
-      rhs=if_exprt(tmp_guard.as_expr(), goto_state_rhs, dest_state_rhs);
+      // tmp_guard-=dest_state.guard;
```
Comment (peterschrammel) [inline src/goto-symex/symex_goto.cpp]: remove

File: src/goto-symex/symex_goto.cpp [hunk 4]
```diff
@@ -366,11 +367,38 @@ void goto_symext::phi_function(
...
+        {
+          assert(tmp_guard2.operands().front()==
+                 goto_state.guard_before_branch);
+          tmp_guard2.operands().erase(tmp_guard2.operands().begin());
+        }
+
+        if(tmp_guard2.operands().size()==1)
+          tmp_guard2=tmp_guard2.op0();
+      }
+      // assert(tmp_guard==tmp_guard2);
```
Comment (peterschrammel) [inline src/goto-symex/symex_goto.cpp]: see above

File: src/solvers/prop/bdd_expr.cpp [hunk 5]
```diff
@@ -165,7 +165,14 @@ exprt bdd_exprt::as_expr(const bdd_exprt::BDDt &r) const
   else if(r.high().is_true())
     return or_exprt(n_expr, as_expr(r.low()));
   else
+    #if 0
```
Comment (peterschrammel) [inline src/solvers/prop/bdd_expr.cpp]: Is there any reason to keep this code?

File: src/util/guard.cpp [hunk 6]
```diff
@@ -134,8 +137,10 @@ Function: operator -=
 
 \*******************************************************************/
 
-guardt &operator -= (guardt &g1, const guardt &g2)
+guardt& guardt::operator -= (const guardt &g2)
```
Comment (peterschrammel) [inline src/util/guard.cpp]: cpplint

File: src/util/guard.h [hunk 7]
```diff
@@ -57,8 +59,8 @@ class guardt:public exprt
   void make_false();
 #endif
 
-  friend guardt &operator -= (guardt &g1, const guardt &g2);
-  friend guardt &operator |= (guardt &g1, const guardt &g2);
+  guardt& operator -= (const guardt &g2);
```
Comment (peterschrammel) [inline src/util/guard.h]: cpplint

File: src/util/simplify_expr_boolean.cpp [hunk 8]
```diff
@@ -87,6 +90,21 @@ bool simplify_exprt::simplify_boolean(exprt &expr)
   {
     if(operands.empty()) return true;
 
+    #if 0
```
Comment (peterschrammel) [inline src/util/simplify_expr_boolean.cpp]: Maybe add a comment why one version is preferred over the other

**Comment by tautschnig** [comment]
Closing as we now have full support for guards as BDDs.

**Pull request closed without merging**